### PR TITLE
Tests grpc public api

### PR DESCRIPTION
**Pull request by modship**
* [ ] document all added functions
* [ ] try in sandbox /simulation/labnet
  * [ ] if part of node-launch, checked using the `resync_check` flag
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [x] if the API has changed, update the API specification

:no_entry_sign: `send_blocks` has been disabled

Code coverage :
image

### CODE REVIEW

File: massa-grpc/src/tests/mock.rs [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+        fn clone_box(&self) -> Box<dyn SelectorController>;
+
+        #[cfg(feature = "testing")]
+        fn get_entire_selection(&self) -> std::collections::VecDeque<(u64,HashMap<Slot,Selection>)> {
+            unimplemented!("mock implementation only")
+        }
+    }
+
+}
+
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/mock.rs:204]: Generate a small doc is better

File: massa-grpc/src/tests/mock.rs [hunk 2]
```diff
@@ -0,0 +1,326 @@
+use std::collections::{BTreeMap, HashMap};
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/mock.rs:1]: Add Massa header

File: massa-grpc/src/tests/stream.rs [hunk 3]
```diff
@@ -0,0 +1,1580 @@
+use std::{net::SocketAddr, ops::Add, time::Duration};
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/stream.rs:1]: Add Massa header

File: massa-grpc/src/tests/stream.rs [hunk 4]
```diff
@@ -0,0 +1,1580 @@
...
+use massa_time::MassaTime;
+use serial_test::serial;
+use tokio_stream::StreamExt;
+
+// const GRPC_SERVER_URL: &str = "grpc://localhost:8888";
+
+#[tokio::test]
+#[serial]
+async fn transactions_throughput_stream() {
+    let addr: SocketAddr = "[::]:4017".parse().unwrap();
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/stream.rs:36]: In other language, they always radomise the ports

File: massa-grpc/src/tests/stream.rs [hunk 5]
```diff
@@ -0,0 +1,1580 @@
...
+        create_operation_with_expire_period,
+    },
+    MockProtocolController,
+};
+use massa_serialization::Serializer;
+use massa_signature::KeyPair;
+use massa_time::MassaTime;
+use serial_test::serial;
+use tokio_stream::StreamExt;
+
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/stream.rs:30]: Try to keep the same function order as the streams in the code(ASC)

File: massa-grpc/src/tests/stream.rs [hunk 6]
```diff
@@ -0,0 +1,1580 @@
...
+#[tokio::test]
+#[serial]
+async fn transactions_throughput_stream() {
+    let addr: SocketAddr = "[::]:4017".parse().unwrap();
+    let mut public_server = grpc_public_service(&addr);
+    let config = public_server.grpc_config.clone();
+
+    let mut exec_ctrl = MockExecutionCtrl::new();
+
+    exec_ctrl.expect_clone().returning(|| {
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/stream.rs:42]: Try to make a small comments in each test to make easily readable: GivenWhenThen

File: massa-grpc/src/tests/stream.rs [hunk 7]
```diff
@@ -0,0 +1,1580 @@
...
+                .unwrap();
+        }
+    }
+
+    stop_handle.stop();
+}
+
+#[tokio::test]
+#[serial]
+async fn new_operations() {
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/stream.rs:155]: I'm not a big fan of long tests, each test should test one behaviour. TO bootstrap your tests, don't hesite to create `helper.rs` file to hide Given test data and helper functions.

File: massa-grpc/src/tests/stream.rs (hunk 4 above)
Comment (modship) [reply massa-grpc/src/tests/stream.rs:36]: i think now i can remove #serial

File: massa-grpc/src/tests/stream.rs [hunk 8]
```diff
@@ -0,0 +1,1580 @@
...
+    public_server.pool_channels.endorsement_sender = endorsement_tx.clone();
+
+    let stop_handle = public_server.serve(&config).await.unwrap();
+
+    let mut public_client = PublicServiceClient::connect(format!(
+        "grpc://localhost:{}",
+        addr.to_string().split(':').into_iter().last().unwrap()
+    ))
+    .await
+    .unwrap();
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/stream.rs:663]: Lot of Given code could be hidden in `helper.rs`, the tests will be lite and far more readable

File: massa-grpc/src/tests/stream.rs [hunk 9]
```diff
@@ -0,0 +1,1580 @@
...
+
+    let endorsement = create_endorsement();
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/stream.rs:665]: It's already done here, I don't went to know how an endorsement is created. Same thing for gRPC client creation

File: massa-grpc/src/tests/stream.rs [hunk 10]
```diff
@@ -0,0 +1,1580 @@
...
+        })
+        .await
+        .unwrap();
+
+    tokio::time::sleep(Duration::from_millis(50)).await;
+
+    endorsement_tx.send(endorsement.clone()).unwrap();
+
+    let result = tokio::time::timeout(Duration::from_secs(2), resp_stream.next()).await;
+    assert!(result.is_err());
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/stream.rs:700]: If we split the big tests, we could separate good/bad scenario tests

File: massa-grpc/src/tests/stream.rs [hunk 11]
```diff
@@ -0,0 +1,1580 @@
...
+        }
+        _ => panic!("should be error"),
+    }
+
+    stop_handle.stop();
+}
+
+#[tokio::test]
+#[serial]
+async fn send_blocks() {
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/stream.rs:1465]: We should disable this one ? as we are planning to disable the `SendBlocks`

File: massa-grpc/src/tests/public.rs [hunk 12]
```diff
@@ -0,0 +1,1707 @@
...
+//     let config = public_server.grpc_config.clone();
+
+//     let stop_handle = public_server.serve(&config).await.unwrap();
+//     let mut public_client = PublicServiceClient::connect(GRPC_SERVER_URL).await.unwrap();
+
+//     stop_handle.stop();
+// }
+#[tokio::test]
+#[serial]
+async fn test_start_grpc_server() {
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/public.rs:81]: It's a king of integration test. We could moveIt to it's own file.

File: massa-grpc/src/tests/stream.rs (hunk 11 above)
Comment (modship) [reply massa-grpc/src/tests/stream.rs:1465]: No, I'm testing that it's unavailable

File: massa-grpc/src/tests/public.rs [hunk 13]
```diff
@@ -0,0 +1,1707 @@
...
+    let mut res = PublicServiceClient::new(channel);
+
+    let s = res.get_status(GetStatusRequest {}).await.unwrap();
+    dbg!(s);
+    stop_handle.stop();
+    std::thread::sleep(Duration::from_millis(1000));
+}
+
+#[tokio::test]
+#[serial]
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/public.rs:220]: Remove serial could boost our test execution, we should find a workaround

File: massa-grpc/src/tests/public.rs [hunk 14]
```diff
@@ -0,0 +1,1707 @@
...
+        addr.to_string().split(':').into_iter().last().unwrap()
+    ))
+    .await
+    .unwrap();
+    let response = public_client.get_status(GetStatusRequest {}).await.unwrap();
+    let result = response.into_inner();
+    assert_eq!(result.status.unwrap().version, *VERSION.to_string());
+
+    stop_handle.stop();
+    std::thread::sleep(Duration::from_millis(1000));
```
Comment (aoudiamoncef) [inline massa-grpc/src/tests/public.rs:251]: @sydhds will not validate this peace of code... We should find a workaround. A helper function which shutdown a server properly.

**Review by aoudiamoncef — CHANGES_REQUESTED** [review]
In general, the PR did what It should done. GG

If we refactor the code little bit, we'll have a great tests

**Review by aoudiamoncef — CHANGES_REQUESTED** [review]
In general, the PR did what It should done. GG

If we refactor the code little bit, we'll have a great tests

File: massa-grpc/src/tests/public.rs (hunk 14 above)
Comment (modship) [reply massa-grpc/src/tests/public.rs:251]: it can be removed

File: massa-grpc/src/tests/public.rs (hunk 13 above)
Comment (modship) [reply massa-grpc/src/tests/public.rs:220]: done

File: massa-grpc/src/tests/stream.rs (hunk 10 above)
Comment (modship) [reply massa-grpc/src/tests/stream.rs:700]: Sometimes i want to verify that it still works after receiving an error

File: massa-grpc/src/tests/stream.rs (hunk 8 above)
Comment (modship) [reply massa-grpc/src/tests/stream.rs:663]: i have mock.rs file for that but i need the server generated in each test to mock them for each case

**Review by aoudiamoncef — APPROVED** [review]
LGTM, great work 👍🏻

**Comment by modship** [comment]
bors merge

**Pull request merged**